sw/host/opentitanlib/src/transport/ti50emulator/emu.rs

```diff
@@ -214,58 +233,54 @@ impl EmulatorProcess {
-            let mut retry = 0;
-            for stream in control_socket.incoming() {
-                match stream {
-                    Ok(mut socket) => {
-                        let len = socket.read(&mut buffer[..]).unwrap();
-                        if PATTERN[..] == buffer[0..PATTERN.len()] {
-                            log::info!("Ti50Emulator ready");
-                        }
-                        return Ok(len);
-                    }
-                    Err(err) if err.kind() == ErrorKind::WouldBlock => {
-                        log::debug!("Wait for sub-process...");
-                        std::thread::sleep(TIMEOUT);
-                        retry += 1;
-                        if retry >= MAX_RETRY {
-                            return Err(EmuError::StartFailureCause(
-                                "Spawning Ti50Emulator sub-process timeout".to_string(),
-                            ));
-                        }
-                    }
-                    Err(_err) => {
-                        return Err(EmuError::RuntimeError(
-                            "Control socket io error".to_string(),
-                        ));
-                    }
-                }
-            }
-            Err(EmuError::StartFailureCause(
-                "Waiting for sub-process failed".to_string(),
-            ))
-        });
-
         log::info!("Spawning Ti50Emulator sub-process");
         log::info!("Command: {} {:?}", exec.display(), args_list);
-        let handle = Command::new(&exec)
+        let mut handle = Command::new(&exec)
             .args(args_list)
+            .stdin(Stdio::null())
+            .stdout(Stdio::piped())
             .spawn()
             .context("Could not spawn sub-process")?;
-        self.proc = Some(handle);
-        match ready_handle
-            .join()
-            .expect("Can't join control socket thread")
-        {
-            Ok(_) => Ok(()),
-            Err(err) => Err(err.into()),
+
+        log::info!("Waiting for sub-process start");
+        let child_stdout = handle.stdout.as_mut().expect("Failed to open stdout");
+        let mut buffer = Vec::new();
+        let mut timeout = TimeVal::microseconds(SPAWN_TIMEOUT.as_micros().try_into()?);
+        let mut fdset = select::FdSet::new();
+        fdset.insert(child_stdout.as_raw_fd());
+        loop {
+            // Select will subtract from the timeout, so that total time waited across all
+            // iterations of the loop is bounded.
+            if select::select(None, Some(&mut fdset), None, None, Some(&mut timeout))? == 0 {
+                // Timeout
+                return Err(EmuError::StartFailureCause(
+                    "Spawning Ti50Emulator sub-process timeout".to_string(),
+                )
+                .into());
+            }
+            let prev_len = buffer.len();
+            buffer.resize(prev_len + 256, 0u8);
+            let read_count = child_stdout.read(&mut buffer[prev_len..])?;
```

Comment:
I think this can be replaced with `read_count = child_stdout.read_to_end(&mut buffer)`.

Reply:
Since you're using `select`, aren't your fds in non-blocking mode?  `read_to_end` will return if the underlying read returns `Ok(0)`.

Reply:
I am not sure if the fd is in non-blocking mode, I read from it once after each time select() has told me that data is ready, meaning that it makes no difference which mode it is in.

If however, I ensured the fd was non-blocking, and `read_to_end()` would return after getting the equivalent of either the kernel returning 0 bytes, or the kernel returning a "would block" error, then I am back at the situation that I need to repeatedly call `read_to_end()` until I encounter the line I want, or give up.  This also means that I would need to carry around what I have read so far, meaning that the parameter to `read_to_end()` would have to be a slice like `&mut buffer[prev_len..]`.

Am I missing something?

Reply:
`read_to_end` takes a `&mut Vec` and always appends, so you could dispense with any slice notation there.  You could also eliminate the subsequent `buffer.resize()` because you'd never have to truncate the excess space after the read.

But, if your fds are blocking, then `read_to_end` is probably not going to behave as you want.

Reply:
That is a fine point, that I initially missed.  Appending to a Vec is exactly what I need.  I can look into making the pipe non-blocking.